pkg/proxy/handlers/spacelister_list_test.go

```diff
@@ -22,127 +22,207 @@ import (
-			},
-			"movielover lists spaces": {
-				username: "movie.lover",
-				expectedWs: []toolchainv1alpha1.Workspace{
-					workspaceFor(t, fakeClient, "movielover", "admin", true),
-				},
-				expectedErr: "",
-			},
-			"signup has no compliant username set": {
-				username:        "racing.lover",
-				expectedWs:      []toolchainv1alpha1.Workspace{},
-				expectedErr:     "",
-				expectedErrCode: 200,
-			},
-			"space not initialized yet": {
-				username:        "panda.lover",
-				expectedWs:      []toolchainv1alpha1.Workspace{},
-				expectedErr:     "",
-				expectedErrCode: 200,
+				}
 			},
-			"no spaces found": {
-				username:        "user.nospace",
-				expectedWs:      []toolchainv1alpha1.Workspace{},
-				expectedErr:     "",
-				expectedErrCode: 200,
+			publicViewerEnabled: true,
+		},
+		"dancelover lists spaces with public-viewer disabled": {
+			username: "dance.lover",
+			expectedWorkspaces: func(fakeClient *test.FakeClient) []toolchainv1alpha1.Workspace {
+				return []toolchainv1alpha1.Workspace{
+					workspaceFor(t, fakeClient, "dancelover", "admin", true),
+					workspaceFor(t, fakeClient, "movielover", "other", false),
+				}
 			},
-			"informer error": {
-				username:        "dance.lover",
-				expectedWs:      []toolchainv1alpha1.Workspace{},
-				expectedErr:     "list spacebindings error",
-				expectedErrCode: 500,
-				overrideInformerFunc: func() service.InformerService {
-					inf := fake.NewFakeInformer()
-					inf.ListSpaceBindingFunc = func(_ ...labels.Requirement) ([]toolchainv1alpha1.SpaceBinding, error) {
-						return nil, fmt.Errorf("list spacebindings error")
-					}
-					return inf
+			publicViewerEnabled: false,
+		},
+	}
+
+	for k, tc := range tests {
+		fakeSignupService, fakeClient := buildSpaceListerFakes(t, tc.publicViewerEnabled)
+
+		t.Run(k, func(t *testing.T) {
+			// given
+			signupProvider := fakeSignupService.GetSignupFromInformer
+			informerFunc := fake.GetInformerService(fakeClient)
+
+			proxyMetrics := metrics.NewProxyMetrics(prometheus.NewRegistry())
+
+			s := &handlers.SpaceLister{
+				GetSignupFunc:          signupProvider,
+				GetInformerServiceFunc: informerFunc,
+				ProxyMetrics:           proxyMetrics,
+			}
+
+			e := echo.New()
+			req := httptest.NewRequest(http.MethodGet, "/", strings.NewReader(""))
+			rec := httptest.NewRecorder()
+			ctx := e.NewContext(req, rec)
+			ctx.Set(rcontext.UsernameKey, tc.username)
+			ctx.Set(rcontext.RequestReceivedTime, time.Now())
+
+			// when
+			ctx.Set(rcontext.PublicViewerEnabled, tc.publicViewerEnabled)
+			ww, err := handlers.ListUserWorkspaces(ctx, s)
+
+			// then
+			require.NoError(t, err)
+			// list workspace case
+			expectedWs := tc.expectedWorkspaces(fakeClient)
+			require.Equal(t, len(expectedWs), len(ww))
+			for i, w := range ww {
+				assert.Equal(t, expectedWs[i].Name, w.Name)
+				assert.Equal(t, expectedWs[i].Status, w.Status)
+			}
+		})
+	}
+}
+
+func TestSpaceListerList(t *testing.T) {
```

Comment:
should we rename this one `TestHandleSpaceListRequest` since that is the function that is invoking ? 

I'm saying this because I was a bit confused at first with the two tests, both testing the publicViewer paths.

Reply:
good catch, done in 2f018f851de0b6e0cf9bac9c456207b4e5932cf1